### DISCUSSION TITLE
Object.freeze 에 대해서 어떻게 생각하시나요 ?

### DISCUSSION THREAD

**Discussion by moonheekim0118** [woowacourse-study/2022-code-review-study#41] [Youtube] [answered]
> Object.freeze() 메서드는 객체를 동결합니다. 동결된 객체는 더 이상 변경될 수 없습니다. 즉, 동결된 객체는 새로운 속성을 추가하거나 존재하는 속성을 제거하는 것을 방지하며 존재하는 속성의 불변성, 설정 가능성(configurability), 작성 가능성이 변경되는 것을 방지하고, 존재하는 속성의 값이 변경되는 것도 방지합니다. 

즉..객체를 동결시켜서 개발자가 객체를 변경하지 못하게 막는건데요, 종종 상수화에 아래처럼 Object.freeze 를 쓰는 크루들이 있었어요. (저도 이번에 썼습니다.)

```javascript
export const ERROR_MESSAGES = Object.freeze({
  EXCEED_MAX_SAVE_AMOUNT: `저장된 비디오의 개수가 ${MAX_SAVE_AMOUNT}개를 초과했습니다.`,
  SERVER_MALFUNCTION: '서버에서 에러가 발생했습니다. 잠시 후 다시 시도하십시오.',
});
```
하지만 이에 관해서 몇가지 의견들이 충돌되어서 다른 크루분들은 어떻게 생각하는지 궁금합니다.

일단 저의 리뷰어님과 제 페어의 리뷰어님은 아래와 같이 코멘트 해주셨어요
> 추가로 Object.freeze 하시는 것도 필요할까 의문이 들어용ㅎㅎ.. 해당 객체에 접근할 수 있는 건 우리 개발자들 밖에 없는데, 변수명부터 나 상수야!!!!!! 건들지마!!! 라서 ㅎㅎㅋㅋㅋㅋ

```
리뷰어님 : 상수에 Object.freeze를 쓰지 않으면 협업시 다른 사람들이 해당 변수의 내용을 바꾸려는 시도를 하게 될까요?
콤피님은 나중에 다른 코드상에서 이 변수의 내용을 바꿀 생각이 들 것 같나요?

페어 : 먼저 저의 의도는 Object는 참조값이니 만약의 상황에서 참조로 인해
혹은 기타 실수로 인해 값이 수정되는 경우를 방지하기 위함이였습니다.

하지만, 변수 이름만 보아도 이것은 상수이다! 라고 말해주어
실제로 변경될 케이스는 확실히 없어보여서, 오버 스팩인 것 같아보여... 로이님의 말씀에도 엄청 공감됩니다.

그런데 제가 과거에 개발자들 간의 언더바를 통한 Private 처리 약속을
지금은 더 이상 사용하지 않는 이유를 찾아보았을 때, "약속일뿐 실제로 변경이 안되는 것은 아니다"라는 문장을 생각해보았을 때

안 써도 되는 것인지에 대해 다시 한번 궁금증이 생기는 것도 사실이라...
이 부분을 어떻게 해야할지 아직까지 고민입니다 😭

실제 현업에서의 Object.freeze()의 용도가 무엇인지 간단한 상황이나 예시에 대해 여쭤보아도 될까요!?

리뷰어님 : "약속일 뿐 실제로 변경이 안되는 것은 아니다" -> 참입니다.
"실제로 변경한다면 그 사람이 약속을 안지킨 것이다" -> 역시 참이죠.
그렇다면 다음과 같이 정리가 될 것 같아요.

누군가 약속을 지키지 않을 것이 염려되는 상황인 경우 -> freeze를 하여 강제성 부여.
팀 내 컨벤션이 확고하여 약속을 지키지 않을 가능성이 현저히 적은 경우 -> 굳이 freeze를 할 필요까지는 없을 듯. 
```

사실 저도 원래 리뷰어님들처럼 'object.freeze 굳이 개발자들끼리 사용하는데 써야하나?' 라고 생각했어요~!
근데 제 페어가 '그런 논리면 private 도 쓸 필요 없는걸까요?' 라고 말해주어서 설득되었습니다....
private 을 사용하는 것과 Object.freeze 를 사용하는것.. 모두 개발자들 끼리 실수를 방지하기 위함이라는 맥락은 같은데
뭐는 써야하고 뭐는 안써야 할까요? @_@ 머리가 핑핑도네요 
여러분은 어떻게 생각하시나요 ?!?

**Comment by byhhh2** [comment, marked as answer]
리뷰어님의 생각대로 `실수할 가능성이 얼마나 있냐?`에 초점에 두는 것이 맞는 것 같습니다. 

> 상수에 Object.freeze를 쓰지 않으면 협업시 다른 사람들이 해당 변수의 내용을 바꾸려는 시도를 하게 될까요?

상수는 이름이 대문자로 되어 있기 때문에 바꾸려고 시도하는 사람의 확률이 현저히 적을 것 같아서요.
